### ISSUE TITLE
WebAuthenticationBroker API that supports all WinAppSDK OS's and application types

### ISSUE THREAD

**Issue by BenJKuhn** [microsoft/WindowsAppSDK#441]
# Proposal: Lifted WebAuthenticationBroker API that supports all Reunion OS's and application types
See #398, #366 - This issue is similar, but distinct in that it covers creating a new, lifted API, rather than updating the existing one to work for more application types.

## Summary
Create an API that developers can use in any application type to implement OAuth2 functionality in any app supported by Reunion. 

## Rationale
The existing OS API that provides this functionality requires a core Window, which, for practical purposes, means that it only works for UWP-style application, and can't be used in other types of applications. Even if the OS API were updated, it would likely only be updated for new OS versions. It would be helpful for developers to have consistent support across OS versions.

https://task.ms/31649739

**Comment by mrlacey** [comment]
And, of course, this should build on top of WebView2!

_Note. The UWP one is based on IE and because of this will not work with many websites, forcing the developer to not be able to use WAB for Oauth with those sites._

**Comment by michael-hawker** [comment]
This was originally a request on the WebView2 team here: https://github.com/MicrosoftEdge/WebView2Feedback/issues/171

There's probably some interop involved I believe between Edge, the OS, and WebView2?

**Comment by dotMorten** [comment]
> And, of course, this should build on top of WebView2!

I disagree. I'd rather we use the user's default browser to log in, then use protocol activation to come back to the application, as it follows better practices, and is impossible to spoof, and also has the benefit of supporting any saved credentials the user might have in their browser. I managed to actually make this work, but it was a lot of work, lots of code, manifest changes for protocol activation, dealing with app activation args, finding the right instance, shutting down the app if it was the wrong instance, redirecting to the correct instance, then dealing with resuming the authentication workflow there. It was quite non-trivial, and I really wish this could be wayyyyyyyyyyy easier.
I wish there was just a single call I could make that would do all this for me. Something like
`public static Task<Uri> OAuthSignIn(Uri authorizationUri, Uri callbackUri);` and protocol activation, app instance tracking and resuming etc is just magically handled for me. I shouldn't have to configure my app at all, just configure my oauth with the right callback uri, and then pass it in here.

**Comment by jonwis** [comment]
> app instance tracking and resuming etc is just magically handled for me. I shouldn't have to configure my app at all, just configure my oauth with the right callback uri, and then pass it in here

I agree with everything you've got here - it's important to use the same browser as the user is used to because it may have SSO cookies already set up making the auth basically no-op, or the user might have a password manager associated with it (Edge, LastPass, etc.) where they store their cookies.

For reference, there is https://github.com/Microsoft/cpprestsdk/blob/master/Release/src/http/oauth/oauth2.cpp (and its friend https://github.com/Microsoft/cpprestsdk/blob/master/Release/include/cpprest/oauth2.h), which we'd have to review to see if it behaves correctly.  VSCode already does this as well, setting itself up a temporary HTTP server to support the redirect, etc.

**Comment by dotMorten** [comment]
Clever trick with the temporary http server. Very clever actually.

**Comment by bgavrilMS** [comment]
If you need a reference implementation, please feel free to use MSAL's, it is MIT licensed. 

[WebView2](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/src/client/Microsoft.Identity.Client/Platforms/Features/WebView2WebUi/WebView2WebUi.cs#L31)
[WebView1](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/src/client/Microsoft.Identity.Client/Platforms/Features/WinFormsLegacyWebUi/WebUI.cs#L24)
[System Browser using http listner](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/src/client/Microsoft.Identity.Client/Platforms/Features/DefaultOSBrowser/DefaultOsBrowserWebUi.cs#L56)

The discussion around embedded vs system browser is long and inconclusive. When it comes to Azure Active Directory, we've concluded that on desktop it's better to use a broker (WAM on Windows) because brokers increase security and SSO. Internally the broker uses an embedded browser. 

Many apps use browsers today and generally the embedded browser are preferred because they provide a better user experience. System browsers are cross-plat but the experience is worse. I have not yet seen any conclusive evidence that system browser is better or worse than embedded browser in terms of security. In terms of SSO, embedded browsers fare poorer, because you are likely to have SSO cookies in your system browser already. But once you start using the embedded browser, it should store the cookies as well. 

Hope this helps.

**Comment by dotMorten** [comment]
Until the WinAppSDK adds this, I've added this ability to WinUIEx v1.4. It's nothing more than a simple one-line call. See https://dotmorten.github.io/WinUIEx/concepts/WebAuthenticator.html

**Comment by mattleibow** [comment]
The solution by @dotMorten works well if you are controlling the app, but as a library - and especially as maui which is a framework that libraries will be based on.

The issue is that the auth callback starts a new process - which may be unexpected for the app, and then it also modifies app state unexpectedly (that may override or be overridden by this): https://github.com/dotMorten/WinUIEx/blob/main/src/WinUIEx/WebAuthenticator.cs#L108

The killing of the new process may also be unexpected for the app:
https://github.com/dotMorten/WinUIEx/blob/main/src/WinUIEx/WebAuthenticator.cs#L100

Not sure what the best solution is, but having this in a popup that is controlled by wasdk is more contained and does not require any new processes.

This pop out to a browser works on Android/iOS because even though a "new window" happens, it is the same process and can be intercepted in known places. iOS has a special place, and Android can have a special activity designed for handling callbacks.

Windows on the other hand launches a brand new app.

I appreciate that popping out to the browser is a "better" scenario from the user's credential perspective and default browser, but this is not super windows-friendly. The callback triggers a new proc, which is killed and even if there was a way to share data, Windows doesn't have a place to decide where a new app should be launched or not.

**Comment by dotMorten** [comment]
>  it also modifies app state unexpectedly

Note: This only happens if the user didn't already set that state.

Personally I'd like to see something that's a combo of a popup and using the default browser - ie some way a browser knows the request is coming from a specific app and knows to direct you back to that specific app instance and close that tab gain, instead of asking the user to open the app again from the browser - ie overall more akin to what iOS and Android does.

**Comment by VR-Architect** [comment]
Any movement or guidance on starting new Maui/WebAPI/BlazorServer combined solution? Is a solution soon?

I am following this Microsoft tutorial but says Windows not working yet.

https://docs.microsoft.com/en-us/dotnet/maui/platform-integration/communication/authentication?tabs=windows

**Comment by VR-Architect** [comment]
Will there be a way to leverage the ASP. NET CORE API server as the backend server? All of my various clients, including Blazor Server, already go through the API for DB access. If yes, any chance someone could make a tuturial for this solution setup?

Our goal is 100% of business logic and per-method authorization is contained in the API so our customers could build their own clients if they choose as we are a SaaS provider.

**Comment by hansmbakker** [comment]
@BenJKuhn is there any update to this? I see @mattleibow added maui-7.x and Maui 7.0 went GA - is there any roadmap or expected moment we can use this?

**Comment by cricketthomas** [comment]
Any update on this issue? **It would be really great for MAUI to have this.**
@davidortinau can this be prioritized at all?

**Comment by sb34000** [comment]
Any update ?

**Comment by Quaybe** [comment]
Please?

**Comment by MatthewBacon-Neteo** [comment]
The fact that this has taken so long to get any traction is incredible.

 Authentication for cross platform MAUI apps is dependent on this functionality and people are literally having to beg for updates.

**Comment by MUN1Z** [comment]
Any updates?

**Comment by kfrancis** [comment]
Any updates?

**Comment by Murtaya** [comment]
Any updated?  We're evaluating using MAUI and....well, are kinda blocked on this and surprised at the lack of responses...

**Comment by cricketthomas** [comment]
> Any update?  We're evaluating using MAUI and....well, are kinda blocked on this and surprised at the lack of responses...

@Murtaya A work around would be to use MSAL/Azure B2C on windows, iOS and Android and WebAuthenticator on MacOS, or WebAuthenticator on everything else, and MSAL on windows

**Comment by matt-goldman** [comment]
A better option is to use WinUIEx as mentioned earlier in this thread

**Comment by emillium** [comment]
I have actually moved completely away from the Microsoft stack because of this.

**Comment by KaleV-K** [comment]
@dotMorten do you have any comment on the progress of this issue??

**Comment by AviNessimian** [comment]
Any update ?

**Comment by MatthewBacon-Neteo** [comment]
@BenJKuhn sorry to at you but is there any progress or words of wisdom with how we can prioritise? 

This is a real put off for many wanting to use Maui for cross platform apps.

**Comment by ronanj2** [comment]
I see a lot of people in this thread looking for updates ... and NOBODY getting back with any.
What is going on?

**Comment by nterreaux** [comment]
I'd like to try the solution with WinUIEx, but I'm using .net 7 and it doesn't seem to be compatible. Is there a solution?

**Comment by dotMorten** [comment]
@nterreaux .net 7 is fully supported by WinUIEx

**Comment by nterreaux** [comment]
I'm now using WinUIEx for my OAuth 2 authentication. Everything works perfectly except for the fact that the user's default browser page doesn't close automatically. This isn't extremely annoying, but is there a solution for closing the browser after authentication?

**Comment by dotMorten** [comment]
@nterreaux since this thread is going more and more off topic let’s try and keep the winuiex questions over in the WinUIEx repo

**Comment by ransagy** [comment]
More because we have zero response from any maintainer, tagged or otherwise, on the gameplan of a core feature of a core platform in this cross-platform framework solution.

Anytime now, MS employees..

**Comment by MatthewBacon-Neteo** [comment]
@ransagy Agree the lack of response hinders any confidence in the framework and Maui. Sad to say that my team / project have had to abandon in favor of other stacks.

**Comment by andreasbrostencab** [comment]
I am also very interested in any status info around this.

**Comment by Azizxon** [comment]
Any updates?

**Comment by Hantse** [comment]
It's a basic usage in app, very important, some feedback ? Should be great if dev team focus on fix this.

**Comment by dotMorten** [comment]
Heads up: Google's OAuth is going to start blocking embedded webviews for OAuth: https://developers.google.com/identity/protocols/oauth2/policies#browsers
This makes it even more important to provide an out of the box secure way to sign in with OAuth, since many are currently just falling back to an embedded WebView2.

**Comment by FedericoSMEC** [comment]
> > A better option is to use WinUIEx as mentioned earlier in this thread
> 
> Except it isnt.
> 
> As reported when the site calls back to the app: "Windows on the other hand launches a brand new app."
> 
> It might be a work around for android / ios but doesnt work for Windows apps.
> 
> Maui is supposed to be cross platform and depends on this functionality.

@MatthewBacon-Neteo A bit late to the party here. I managed to use a workaround for windows that might help you. 

When the second instance of the app is initializing (before the window of the second instance is created), I extract the auth code from my custom uri (e.g. mycustomuri://authorization?authCode=abc), I then use Named Pipes to send this code to the first instance of the app (the one that is waiting for the code) and then kill the second instance. 

It is quite a workaround but on Windows it works fine and the second app is not seen by the user. Of course, fixing this issue would make everything easier.

**Comment by dotMorten** [comment]
You don't have to use your own named pipes to redirect. There's an actual API for that (`RedirectActivationToAsync`). See https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/applifecycle/applifecycle-instancing#redirection-logic-based-on-activation-kind

**Comment by FedericoSMEC** [comment]
Thanks @dotMorten. I don't want to derail the conversation, although I fail to see how I can use that RedirectActivationToAsync in my .net maui app.

**Comment by IlGalvo** [comment]
I think I found a workaround after many tries:

```
var redirectUrl = "http://127.0.0.1:{0}/authorize/"; // or "http://localhost:{0}/authorize/"

var tcplistener = new TcpListener(IPAddress.Loopback, 0);
tcplistener.Start();

var port = ((IPEndPoint)tcplistener.LocalEndpoint).Port;
tcplistener.Stop();

redirectUrl = string.Format(redirectUrl, port);

using var httpListener = new HttpListener();
httpListener.Prefixes.Add(redirectUrl);
httpListener.Start();

var oAuth2Url = "your-oauth2-url"; // make sure (if present) redirect_url = redirectUrl
Process.Start(new ProcessStartInfo("cmd", $"/c start \"\" \"{oAuth2Url}\"") { CreateNoWindow = true });

var context = await httpListener.GetContextAsync();

var html = @"<html><head><title>OAuth 2.0 Authentication Token Received</title></head><body> Received verification code. You may now close this window.</body></html>";
var bytes = Encoding.UTF8.GetBytes(html);

context.Response.ContentLength64 = bytes.Length;
context.Response.SendChunked = false;
context.Response.KeepAlive = false;
var output = context.Response.OutputStream;
await output.WriteAsync(bytes);
await output.FlushAsync();
output.Close();
context.Response.Close();

var result = context.Request.QueryString; // contains url result params (eg: code etc)
```

The only encountered problem is: if you close the browser before user-auth-flow completes,
`var context = await httpListener.GetContextAsync();`
hangs indefinitely.

**Comment by elijahmontenegro** [comment]
Has anybody found a workaround for use in MAUI apps? I would appreciate if you can share any insight on this! Really frustrating!

**Comment by illja96** [comment]
@FedericoSMEC can you post your code for a workaround, please?

**Comment by lcarrion2406** [comment]
Any update?

**Comment by dotMorten** [comment]
> Has anybody found a workaround for use in MAUI windows app

@elijahmontenegro  Yes: https://dotmorten.github.io/WinUIEx/concepts/Maui.html#use-winuiexs-webauthenticator-instead-of-net-mauis

**Comment by ataparia** [comment]
> ue is that the auth callback starts a new process - which may be unexpected for the app, and then it also modifies app state unexpectedly (that may override user data or be overridden by user data):

Hi @dotMorten , I tried to reference WinUIEx.WebAuthenticator in the App.xaml.cs but I don't see CheckOAuthRedirectionActivation() method on there. I am using the WinUIEx package as follows:

<PackageReference Include="WinUIEx" Version="1.4.230822000" Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'" />

Am I missins something?

**Comment by dotMorten** [comment]
@ataparia Let's do this over at https://github.com/dotMorten/WinUIEx/discussions

**Comment by dotMorten** [comment]
Looks like not on Google, but Facebook has also banned the use of embedded browsers and asking developers to use platform-provided authentication APIs on Android: https://developers.facebook.com/blog/post/2021/06/28/deprecating-support-fb-login-authentication-android-embedded-browsers/
At some point this'll likely come to Windows too.

**Comment by Brosten** [comment]
I'm also having this issue. The workaround suggested might be an option, but please fix! Issues like this makes developers run away from .NET MAUI. :(

**Comment by Brosten** [comment]
I have tried the workarounds from 
https://dotmorten.github.io/WinUIEx/concepts/Maui.html#use-winuiexs-webauthenticator-instead-of-net-mauis
and from here
https://github.com/dotnet/maui/issues/2702#issuecomment-1172988972

Both seem to do about the same thing. It does fire up a browser with the login page, however, the created url-parametes gives me 403/forbidden on our auth service. 
I can work around this, by using the second workaround and comment out the section manipulating the state parameter in the url. In this case, I'm able to login, but while doing so, my app isn't blocked in the background. I would have hoped this would be the case, or do I need to take care of that myself? 
The bigger problem is that when the browser return it doesn't return to my running app, but instead launches a new instance of the app and returns there... Does this has to do with my manipulation of the state parameter? Or am I doing some other mistake?

**Comment by beto-rodriguez** [comment]
so, it seems that embedded browsers it not the way ([1](https://github.com/microsoft/WindowsAppSDK/issues/441#issuecomment-1764951494), [2](https://github.com/microsoft/WindowsAppSDK/issues/441#issuecomment-1599085930)), does anyone know an article that helps to use the user browser to authenticate the user in a MAUI app?

**Comment by jbt00000** [comment]
This issue is over 2.5 years old.  Is there any reason why the assignee @ujjwalchadha has not commented on this?

**Comment by Mogikan** [comment]
> This issue is over 2.5 years old. Is there any reason why the assignee @ujjwalchadha has not commented on this?

Probably they are busy with Map control, which is older than this feature. At least Maps are in roadmap and this feature is not planned. In best scenario it will be implemented within 5 years.

**Comment by solomonfried** [comment]
Do we know if anyone on the MAUI team is monitoring this issue? This is a critical block for migrating to MAUI. 
My Xamarin app that uses WebAuthenticator, already is blocked by Facebook and Apple on UWP from connecting due to its use of IE11. 
Hard to understand why this is not begin given the highest priority since most serious apps require Authentication.

Between this issue and the inability to publish MAUI UWP apps that reference a MAUI class library ([https://github.com/dotnet/maui/issues/8304](https://github.com/dotnet/maui/issues/8304)), it feels that UWP support for MAUI is being de-prioritized. Wondering if I should drop the platform and only support Android and iOS.

**Comment by dotMorten** [comment]
Last thing we were told (during community standup I believe), was that it isn't planned for v1.5.

**Comment by RobinS-S** [comment]
Can we get some attention on this issue? The problem has existed for quite some time now and prevents simple things such as authentication from working in MAUI for Windows. And I can't imagine MAUI is the only one that uses this.

It's unbelievable how Microsoft is trying to push for a new standard for cross-platform application development but can't manage to make their own OS work with it.

**Comment by ThreeSevenths** [comment]
> Is there any other way to implement social logins in MAUI besides using web authenticator? I tried it today and on windows it throws an exception without an actual exception message. I only want to develop desktop app for windows and mac, no mobile needed and I'm scrathing my head over this.
> 
> ```
> blazor.webview.js:1  
>    at WinRT.ExceptionHelpers.<ThrowExceptionForHR>g__Throw|39_0(Int32 hr)
>    at ABI.Windows.Security.Authentication.Web.IWebAuthenticationBrokerStaticsMethods.AuthenticateAsync(IObjectReference _obj, WebAuthenticationOptions options, Uri requestUri, Uri callbackUri)
>    at Windows.Security.Authentication.Web.WebAuthenticationBroker.AuthenticateAsync(WebAuthenticationOptions options, Uri requestUri, Uri callbackUri)
>    at Microsoft.Maui.Authentication.WebAuthenticatorImplementation.AuthenticateAsync(WebAuthenticatorOptions webAuthenticatorOptions)
>    at MauiHybrid.Components.Pages.Home.Authenticate()
>    at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
>    at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)
> ```

Are you able to implement a protocol handler or run a web server on a port, and do your login via the user's browser? This seems to be the way some apps handle the login problem.

Login in App -> Start web server [optional] -> Open browser to login - > User logs in via browser -> callback url is either https://localhost:10234/login_callback or myawesomeapp://login_callback which gets the token back to you and completes the process.

**Comment by dotMorten** [comment]
@vukasinpetrovic see https://dotmorten.github.io/WinUIEx/concepts/Maui.html#use-winuiexs-webauthenticator-instead-of-net-mauis

**Comment by solomonfried** [comment]
Thank you @dotMorten 
I will try this. 

But the question is, if there is a solution why Microsoft can't build it into WebAuthenticator.
You would think that they would want to support their own platform! Authenticating users is essential to most serious applications.

**Comment by dotMorten** [comment]
@solomonfried Preaching to the choir. I've done my fair share of pushing for this, all the way to the point where I had meetings with the team that promised would look into doing this and explaining the reasonings why it is important, but I haven't seen any progress yet. I'm just as puzzled as you, but they are well aware of the limitation at this point, but I don't know what is keeping the priority of this back - I can't really make sense of it.

On the other hand I have no clue why .NET MAUI has left the old UWP implementation in place for MAUI when it doesn't work outside UWP, rather than just deleting all that code and throwing a more helpful PlatformNotSupported exception. Perhaps @pureween can answer that.

**Comment by mominshaikhdevs** [comment]
@BenJKuhn @codendone this is a crucial API. whats the latest update on this?

**Comment by solomonfried** [comment]
Does anyone know if WebAuthenticator will finally be supported in .NET 9?
Is there an alternative MAUI solution for OAuth? Is there a 3rd party NuGet package? What are all of you doing to manage this?

I am stuck supporting my app on Xamarin because I can't implement a 3rd party login on Windows.

**Comment by dotMorten** [comment]
It's on the 1.7 roadmap

**Comment by michael-hawker** [comment]
🦙 See 1.7 Plans Discussion which calls this out here: https://github.com/microsoft/WindowsAppSDK/discussions/4710

**Comment by solomonfried** [comment]
@michael-hawker, @dotMorten 
Thanks for the update. 
I guess I'll have to wait till 1.7 is released. 
Any estimate on a date?

**Comment by blmiles** [comment]
Hello @dotMorten - just a little more on this issue ...

With the latest .Net 9.0, the WebView is broken. IT no longer triggers the Navigated and Navigating events at all.

These no longer fire at all. Can't find a workaround.
This WAS working in Windows, was not needed in iOS or Android (for all reasons in this whole thread) and now the Windows functionality is fully broken because of this new issue.

I'll take a look at the fix posted earlier but thought I'd mention this.


            _webView.Navigated += (sender, e) =>
            {
                if (e.Url.StartsWith(options.EndUrl))
                {
                    if (tcs.Task.Status != TaskStatus.RanToCompletion)
                    {
                        tcs.SetResult(new BrowserResult
                        {
                            ResultType = BrowserResultType.Success,
                            Response = e.Url.ToString()
                        });
                    }
                }
            };

            _webView.Navigating += (sender, e) =>
            {
                  //do something
            };

**Comment by ghostidentity** [comment]
Can anyone provide a demo project to verify its working targeting 1.7? e.g. an app can connect to google auth2 ?

**Comment by Eddie-Hartman** [comment]
@mattleibow What is going on with this issue and how it relates to https://github.com/dotnet/maui/issues/2702 ?
The above issue has been open for 3+ years at this point and closed. This current issue has been open for 4+ years. How do we upvote it to get it prioritized higher?
My understanding is that this is a blocker for some auth flows for Windows, where I for example want to implement OIDC auth for Windows in .NET MAUI and OIDC library authors have cited not supporting this in part because of these limitations.
The documentation says to follow the above issue, which is marked not doing and then points to this issue.
Can we get an overall update on how this stands and currently affects MAUI and the plan to address it?

**Comment by dotMorten** [comment]
@Eddie-Hartman did you see the 1.7 previews? It has website APIs

**Comment by ghostidentity** [comment]
@mrlacey Here's the alternative. You can still use MSAL utilizing Azure B2C https://www.youtube.com/watch?v=Qyx3Va04dUI or webview2 with Auth0 using PKCE https://youtu.be/SHx364mzkqo?si=3urXUtM9JpBnoXhm  . Though with PKCE, you will need to chain it with protocol activation. https://youtu.be/SHx364mzkqo?si=3urXUtM9JpBnoXhm

**Comment by Eddie-Hartman** [comment]
@dotMorten yes sir. Thank you. I was looking more for how it relates back to MAUI and when that might get released and then implemented. I don't know when 1.7 would be coming out, and then from there when it would be updated in MAUI (if any work needs completed there to implement/update it). FWIW I also saw the workarounds, but trying to get a more unified dev experience in MAUI and so some libraries can get updated as well.

There is quite a lot going on between the two issues and on the MAUI side, comments are closed on the original issue and you can't even like the issue to raise the priority, but it's the one pointed to from the documentation, and then is pointed here. @mattleibow does there need to be a new ticket on the MAUI side since this is getting closer to being able to be implemented or how would you prefer we raise some awareness/priority for this work?

Reading through the roadmap, I guess 25h1 means 1.7 will come out in the first half of 2025? So can we immediately upgrade after that and then this will "just work" or will we have to wait for a MAUI patch?

Thanks.

**Comment by solomonfried** [comment]
I have been waiting for this "to just work" without Win-specific code or other workarounds. 
I am also waiting for 1.7 to finally fix the problem. My hope is that the documentation on WebAuthenticator will no longer have the Warning that 
`At the moment, [WebAuthenticator](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.authentication.webauthenticator) isn't working on Windows`

[](https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/communication/authentication?view=net-maui-9.0&tabs=windows)

For now, I am just proceeding with the iOS/Android version of my app and hope that this will be ready shortly.

**Comment by blmiles** [comment]
> [@mrlacey](https://github.com/mrlacey) Here's the alternative. You can still use MSAL utilizing Azure B2C https://www.youtube.com/watch?v=Qyx3Va04dUI or webview2 with Auth0 using PKCE https://youtu.be/SHx364mzkqo?si=3urXUtM9JpBnoXhm . Though with PKCE, you will need to chain it with protocol activation.

@ghostidentity - I use Auth0 now and no the PKCE
Any sample code for Maui/C# app targeting WinUI 3

Thanks

**Comment by afpa7** [comment]
I opened a discussion in MAUI repo hoping someone would clarify.

https://github.com/dotnet/maui/discussions/27981

**Comment by rahul7720** [comment]
Looks like the issue is fixed in Win [App SDK  Release 1.7.1](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#new-authentication-api)

**Comment by solomonfried** [comment]
> Looks like the issue is fixed in [Win App SDK Release 1.7.1](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#new-authentication-api)

Does this mean that WebAuthenticator now works for UWP?
The documentation does not seem to be updated and still shows a warning message for UWP
https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/communication/authentication?view=net-maui-9.0&tabs=windows

**Comment by dotMorten** [comment]
@solomonfried that link is for Maui which doesn’t use UWP. Wrt Maui they’d still need to do work to adopt the API

**Comment by brockallen** [comment]
> Wrt Maui they’d still need to do work to adopt the API

I don't work/live in the MAUI space so sorry if my follow up is obvious, but can I ask -- so this means there's hope that this will finally be harmonized?

**Comment by solomonfried** [comment]
> [@solomonfried](https://github.com/solomonfried) that link is for Maui which doesn’t use UWP. Wrt Maui they’d still need to do work to adopt the API

My mistake, I meant to refer to the Windows App platform in MAUI. (I have updated my original comment)

**Comment by IlGalvo** [comment]
If anyone needs it, I have created a sample repo for using oauth2manager in winui3 packaged/unpackaged:
https://github.com/IlGalvo/WinUI3-OAuth2Manager-Sample
it is still work-in-progress but I plan to finish it in the next few days, but it seems to be working.

**Comment by theofanis** [comment]
How does one approach this today? With https://github.com/microsoft/WindowsAppSDK/pull/4772 merged, what is the state of this ticket?

Mostly interested in why MAUI still cannot work with this on Windows. The [older issue](https://github.com/dotnet/maui/issues/2702) in the maui repo is now closed, since it was linking to this ticket here.

Looking for a friend. Thank you!

**Comment by solomonfried** [comment]
Is anyone looking into this. Really frustrating that over a year into MAUI, Windows still does not support this basic functionality.
Calling WebAuthenticator results in an exception
`This implementation of WebAuthenticator does not support Windows. See https://github.com/microsoft/WindowsAppSDK/issues/441 for more details.`

Microsoft is still touting WebAuthenticator as the preferred method of implementing OAuth 
[https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/communication/authentication?view=net-maui-9.0&tabs=windows](https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/communication/authentication?view=net-maui-9.0&tabs=windows)

What is the solution? How do we support basic login to OAuth providers (Google, AD, ....)

**Comment by dotMorten** [comment]
@solomonfried not sure why this issue is still open but this api was added a few releases ago. The issue is in MAUIs court and I already started a prototype with initial work in the Maui repo but waiting for the Maui team to help make some decisions on the direction they want to take it.

**Comment by solomonfried** [comment]
@dotMorten 
Thanks for the response. I had also heard that the underlying issues were resolved in the latest Win SDK. I have an education app used by many teachers that have Surface tablets. I was hoping to be converted to MAUI before the new school year starts. 
This lack of support of such a basic requirement, helps fuel the rumors of MAUI being phased back (or out). 
Do I look at third party solutions such as [https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/communication/authentication?view=net-maui-9.0&tabs=windows](https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/communication/authentication?view=net-maui-9.0&tabs=windows)
or do I hope that the MAUI team will get this resolved soon?
Are you aware of any other solutions to this problem?
Any ideas of how to escalate this to someone on the team?
Thanks

**Comment by dotMorten** [comment]
@solomonfried you can just call the oauth APIs in the winappsdk instead of the Maui one on windows or use my WinUIEx library (which works pretty much the same way): https://dotmorten.github.io/WinUIEx/concepts/Maui.html#use-winuiexs-webauthenticator-instead-of-net-mauis

**Comment by solomonfried** [comment]
@dotMorten Thanks for the info. I will definitely be trying that out. Would you have any documentation as to the reason behind the ConfigureLifecycleEvents and the section on "Interact with the Page's containing window:"

Will this result in the same user experience as WebAuthenticator (under Xamarin)?

Thank again.

**Comment by ssparach** [comment]
Thank you all for your engagement on this issue!

We’ve introduced a new API called OAuth2Manager to support web authentication. You can find more details in the [release notes](https://learn.microsoft.com/windows/apps/windows-app-sdk/stable-channel#:~:text=New%20Authentication%20API) and the [documentation](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.security.authentication.oauth.oauth2manager?view=windows-app-sdk-1.7).
If you encounter any bugs or have additional feedback while using this API, please feel free to open a new issue in the repository.

Thanks again for your contributions and support!

**Comment by solomonfried** [comment]
Does  OAuth2Manager resolve the issue of Authentication support in MAUI apps which had been handled by WebAuthenticator, which is still not supported for Windows targets in MAUI?

**Comment by dotMorten** [comment]
@solomonfried no the Maui team still needs to do some significant work here and the api doesn’t make it easy on them. 
Unfortunately the new api winappsdk shipped doesn’t actually solve some of the major hurdles with oauth in windows apps, and provided an api anyone could have built (like what is WinUIEx) including Maui. The workflow with app activation and redirect is just rather problematic and doesn’t allow for an easy application developer story

**Comment by solomonfried** [comment]
@dotMorten thanks for the (discouraging) information.
Is there a general consensus as to what developers should do to handle this. Should we just stop releasing Windows versions of MAUI apps?

**Comment by dotMorten** [comment]
You can totally do oauth in a Maui on windows using these APIs. It’s just a lot more work on you and you can’t use mauis own APIs to do it.

**Comment by solomonfried** [comment]
I guess I'll have to bite the bullet. I just can't understand why if OAuth2Manager will work, why doesn't the WebAuthenticator team use that for the Windows implementation?

**Comment by snow-jallen** [comment]
> You can totally do oauth in a Maui on windows using these APIs. It’s just a lot more work on you and you can’t use mauis own APIs to do it.

Could someone point to a tutorial on _how_ one might "do oauth in a Maui on windows using these APIs"?

**Comment by SongPing** [comment]
This issue is more than 4 years old.
Is there no solution yet? 
This is very disapointing.

**Comment by solomonfried** [comment]
Now December 2025 and still no support. 
I am spending hours trying to get an alternative using WinUIEx or anything else. 

I would love to just drop Windows support for my app, but it is used in schools and many have Surface tablets.

If ANYONE has resolved this, please post.

**Comment by Mogikan** [comment]
My opinion is that Xamarin and WinUI are dead. Don't use them. .NET is not dead only in Web (and maybe WPF, WinForms). I don't like it to happen either, since I developed for it 10 years+, but life is life. Jump off of the dead horse. 100% definition is that it is definitely not yet dead now. But it will.
PS: What I am trying to say is that they are alive (but there are signs that they are low priority projects like running iOS apps on intel desktop, maps and inkcontrols on WinUI, hanging for 5+ years bugs, retired VS mac), but looking at numbers of adoption % and market share (check google trends vs Flutter or React Native and WinUI against WPF, WinForms) and remembering Silverlight and Windows Phone story it is obvious what will happen next.

**Comment by philnwoha** [comment]
I made a post [here](https://github.com/dotMorten/WinUIEx/discussions/263) on how I was able to get this working for my .net maui windows app.

**Comment by solomonfried** [comment]
Thanks for the code. I will look at it. Does it work with .NET 10? Will it work in a few months with ,NET 11?
Should we really give up on Microsoft getting their act together and finally closing the WebAuthentication issue? 
It's so frustrating that this critical MAUI (write once, run anywhere) library, works on all platforms except for Microsoft's own. 
Is there anyone in the Microsoft managerial chain that can get the WinSdk and MAUI teams to get together and finally solve this?

**Comment by philnwoha** [comment]
> Thanks for the code. I will look at it. Does it work with .NET 10? Will it work in a few months with ,NET 11? Should we really give up on Microsoft getting their act together and finally closing the WebAuthentication issue? It's so frustrating that this critical MAUI (write once, run anywhere) library, works on all platforms except for Microsoft's own. Is there anyone in the Microsoft managerial chain that can get the WinSdk and MAUI teams to get together and finally solve this?

@solomonfried I have not tested on .net 10 but I imagine it should work but you may have to fiddle with the windows sdk version as mentioned on the WinUiEx site.  As for the future, it would be fantastic if Microsoft delivered an official solution, but until then this is better than nothing I suppose.

**Comment by dotMorten** [comment]
> Does it work with .NET 10? Will it work in a few months with ,NET 11?

Yes.

**Comment by solomonfried** [comment]
I may have to bite the bullet and implement native WinUI authentication.

I’m a small business selling apps to schools. A few years ago, I always recommended Surface tablets. I developed on them myself, and my apps consistently ran best on Windows.

The version of my app currently in the Microsoft Store is still Xamarin-based. I already have a .NET MAUI version using WebView2, but I’ve held off releasing it because embedded browsers are increasingly discouraged for OAuth authentication. It still works today with providers like Google and Microsoft, but they explicitly reserve the right to stop supporting embedded browsers at any time.

Recently, I signed up a new school, and for the first time I recommended iPads instead.

The reason had nothing to do with hardware. I simply can’t guarantee that authentication on Windows will work as cleanly or reliably as it does on the other platforms. On iOS and Android, authentication is straightforward and well-supported. On Windows, I’m dealing with browser handoffs, platform-specific code, and uncertainty around future support.

That recommendation alone may represent anywhere from a few dozen to a few hundred devices this year. It won’t impact Microsoft’s numbers in any meaningful way, but it makes me wonder how many larger vendors are making the same calculation.

Authentication is the front door to every serious app. It’s astonishing that Microsoft has allowed this situation to persist for so long.
